Objects/exceptions.c

```diff
@@ -3064,44 +3082,50 @@ PyObject *PyExc_UnicodeEncodeError = (PyObject *)&_PyExc_UnicodeEncodeError;
 static int
 UnicodeDecodeError_init(PyObject *self, PyObject *args, PyObject *kwds)
 {
-    PyUnicodeErrorObject *ude;
-
-    if (BaseException_init((PyBaseExceptionObject *)self, args, kwds) == -1)
+    if (BaseException_init((PyBaseExceptionObject *)self, args, kwds) == -1) {
         return -1;
+    }
+
+    PyUnicodeErrorObject *exc = (PyUnicodeErrorObject *)self;
 
-    ude = (PyUnicodeErrorObject *)self;
+    Py_CLEAR(exc->encoding);
+    Py_CLEAR(exc->object);
+    Py_CLEAR(exc->reason);
 
-    Py_CLEAR(ude->encoding);
-    Py_CLEAR(ude->object);
-    Py_CLEAR(ude->reason);
+    PyObject *encoding = NULL, *object = NULL, *reason = NULL;  // borrowed
+    Py_ssize_t start = -1, end = -1;
 
     if (!PyArg_ParseTuple(args, "UOnnU",
-                          &ude->encoding, &ude->object,
-                          &ude->start, &ude->end, &ude->reason)) {
-             ude->encoding = ude->object = ude->reason = NULL;
-             return -1;
+                          &encoding, &object, &start, &end, &reason))
+    {
+        return -1;
     }
 
-    Py_INCREF(ude->encoding);
-    Py_INCREF(ude->object);
-    Py_INCREF(ude->reason);
+    if (start < 0) {
+        PyErr_SetString(PyExc_ValueError, "'start' must be >= 0");
+        return -1;
+    }
 
-    if (!PyBytes_Check(ude->object)) {
+    if (!PyBytes_Check(object)) {
         Py_buffer view;
-        if (PyObject_GetBuffer(ude->object, &view, PyBUF_SIMPLE) != 0)
-            goto error;
-        Py_XSETREF(ude->object, PyBytes_FromStringAndSize(view.buf, view.len));
+        if (PyObject_GetBuffer(object, &view, PyBUF_SIMPLE) != 0) {
+            return -1;
+        }
+        PyObject *content = PyBytes_FromStringAndSize(view.buf, view.len);
         PyBuffer_Release(&view);
-        if (!ude->object)
-            goto error;
+        if (content == NULL) {
+            return -1;
+        }
+        Py_INCREF(object);          // make 'object' a strong reference
+        Py_SETREF(object, content);
     }
```

Comment:
Do you want to modify object or ude->object?

Reply:
I don't want to touch `ude->object` until I'm done with checks and conversion, hence `object` only.

The way I counted my references is as follows:

- `object` is borrowed because of `PyArg_ParseTuple`
- If I call `Py_(X)SETREF(object, something)`, then it's semantically equivalent to `Py_DECREF(object); object = something;`. But since 'object' was borrowed, I first need to incref its reference count and then am I able to call `Py_(X)SETREF(object, something)`. The borrowed reference wouldn't have its count modified so I don't care anymore about it.

Now that I think about, should I also decref `content` just after since I'm doing `Py_NewRef(object)` a few lines below? (I feel I'm actually increasing the reference count of `object` twice in this case...).

Reply:
Thank you!